### ISSUE TITLE
Scope checking for require

### ISSUE THREAD

**Issue by samdark** [vimeo/psalm#11677]
`hello.php`:

```php
<?php

function render(): void {
    require 'template.php';
};

render();
```

`template.php`:

```php
<?php
/** @var string $name */
?>
Hello, my name is <?= htmlentities($name, ENT_QUOTES) ?>!
```

In the example template, `` is required to be defined in the scope of the `require` call, so I'd expect static analysis to detect the missing required scope variable.

_// this is useful for "PHP templates" implementation_

**Comment by samdark** [comment]
No. It requires 2 files.

**Comment by kkmuffme** [comment]
Psalm does detect and report it, but because you used @var you manually overwrote this yourself, disabling the error.

Remove the @var or u disableVarParsing="true" (https://psalm.dev/docs/running_psalm/configuration/#disablevarparsing) in the psalm config file

**Comment by kkmuffme** [comment]
Instead of @var you should be using `assert( isset( $name ) && is_string( $name ) );`

This will ensure that PHP automatically checks that the variable is declared in runtime making it safe if it's changed externally, will allow psalm to detect the type and will not run in production (= no performance penalty) PHP, since assert() statements are noop/ignored there (see the php.ini section on assert for that)

**Comment by samdark** [comment]
OK. But what if it is not a string but a Psalm type?

**Comment by danog** [comment]
Psalm types can be asserted in the same way through assert.

**Comment by samdark** [comment]
I was not aware of that. Is that anywhere in the docs?
